crates/aptos-crypto/src/asymmetric_encryption/elgamal_curve25519_aes256_gcm.rs

```diff
@@ -0,0 +1,156 @@
+        let aes_key_bytes = Self::hash_group_element_to_aes_key(&aes_key_g1.compress());
+        let key = Key::<Aes256Gcm>::from_slice(aes_key_bytes.as_slice());
+        let cipher = Aes256Gcm::new(key);
+        let nonce = Aes256Gcm::generate_nonce(aead_rng);
+        let nonce_bytes = nonce.to_vec();
+        ensure!(
+            12 == nonce_bytes.len(),
+            "ElGamalCurve25519Aes256Gcm enc failed with unexpected nonce len"
+        );
+
+        let aes_ciphertext = cipher.encrypt(&nonce, msg.as_ref()).map_err(|e| {
+            anyhow!(
+                "ElGamalCurve25519Aes256Gcm enc failed with aes error: {}",
+                e
+            )
+        })?;
+
+        let elgamal_ciphertext_0_bytes = elgamal_ciphertext_0.compress().to_bytes().to_vec();
+        let elgamal_ciphertext_1_bytes = elgamal_ciphertext_1.compress().to_bytes().to_vec();
+
+        let serialized = [
+            elgamal_ciphertext_0_bytes, // 32 bytes
+            elgamal_ciphertext_1_bytes, // 32 bytes
+            nonce_bytes,                // 12 bytes
+            aes_ciphertext,             // variable length
+        ]
+        .concat();
+
+        Ok(serialized)
+    }
+
+    fn dec(sk: &Scalar, ciphertext: &[u8]) -> anyhow::Result<Vec<u8>> {
+        ensure!(
+            ciphertext.len() >= 76,
+            "ElGamalCurve25519Aes256Gcm dec failed with invalid ciphertext length"
+        );
+        let c0 = CompressedEdwardsY::from_slice(&ciphertext[0..32])
+            .decompress()
+            .ok_or_else(|| {
+                anyhow!("ElGamalCurve25519Aes256Gcm dec failed with invalid c0 element")
+            })?;
+
+        ensure!(
+            c0.is_torsion_free(),
+            "ElGamalCurve25519Aes256Gcm dec failed with non-prime-order c0"
+        );
+
+        let c1 = CompressedEdwardsY::from_slice(&ciphertext[32..64])
+            .decompress()
+            .ok_or_else(|| {
+                anyhow!("ElGamalCurve25519Aes256Gcm dec failed with invalid c1 element")
+            })?;
+
+        ensure!(
+            c1.is_torsion_free(),
+            "ElGamalCurve25519Aes256Gcm dec failed with non-prime-order c1"
+        );
+
+        let aes_key_element = elgamal::decrypt::<Curve25519>(sk, &c0, &c1).compress();
+        let aes_key_bytes = Self::hash_group_element_to_aes_key(&aes_key_element);
+        let key = Key::<Aes256Gcm>::from_slice(aes_key_bytes.as_slice());
+        let cipher = Aes256Gcm::new(key);
+        let nonce = Nonce::<Aes256Gcm>::from_slice(&ciphertext[64..76]);
+        let plaintext = cipher.decrypt(nonce, &ciphertext[76..]).map_err(|e| {
+            anyhow!("ElGamalCurve25519Aes256Gcm dec failed with aes decryption error: {e}")
+        })?;
+        Ok(plaintext)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::asymmetric_encryption::{
+        elgamal_curve25519_aes256_gcm::ElGamalCurve25519Aes256Gcm, AsymmetricEncryption,
+    };
+
+    #[test]
+    fn gen_enc_dec() {
```

Comment:
I would add tests for encrypting & decrypting messages of length 0.

Reply:
fixed